src/server/coprocessor/endpoint.rs

```diff
@@ -315,12 +271,216 @@ fn is_point(range: &KeyRange) -> bool {
+                        Some(bs) => {
+                            if self.aggr_cnt > 0 {
+                                data.push(box_try!(bs.as_slice().decode_datum()));
+                            } else {
+                                if !data.is_empty() {
+                                    box_try!(datum::encode_to(&mut encoded, &data, false));
+                                    data.clear();
+                                }
+                                encoded.extend(bs);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        if self.aggr_cnt > 0 {
+            try!(self.aggregate(data));
+            return Ok(None);
+        }
+        let mut row = Row::new();
+        let handle = box_try!(datum::encode_value(&[Datum::I64(h)]));
+        row.set_handle(handle);
+        if !data.is_empty() {
+            box_try!(datum::encode_to(&mut encoded, &data, false));
+        }
+        row.set_data(encoded);
+        Ok(Some(row))
+    }
+
+    fn get_group_key(&mut self) -> Result<Vec<u8>> {
+        let items = self.sel.get_group_by();
+        if items.is_empty() {
+            return Ok(SINGLE_GROUP.to_vec());
+        }
+        let mut vals = Vec::with_capacity(items.len());
+        for item in items {
+            let v = box_try!(self.eval.eval(item.get_expr()));
+            vals.push(v);
+        }
+        let res = box_try!(datum::encode_value(&vals));
+        Ok(res)
+    }
+
+    fn aggregate(&mut self, row: Vec<Datum>) -> Result<()> {
+        for (col, d) in self.sel.get_table_info().get_columns().iter().zip(row) {
+            self.eval.row.insert(col.get_column_id(), d);
+        }
+        let gk = Rc::new(try!(self.get_group_key()));
+        let aggr_exprs = self.sel.get_aggregates();
+        match self.gk_aggrs.entry(gk.clone()) {
+            Entry::Occupied(e) => {
+                let funcs = e.into_mut();
+                for (expr, func) in aggr_exprs.iter().zip(funcs) {
+                    // TODO: cache args
+                    let args = box_try!(self.eval.batch_eval(expr.get_children()));
+                    try!(func.update(&args));
+                }
+            }
+            Entry::Vacant(e) => {
+                let mut aggrs = Vec::with_capacity(aggr_exprs.len());
+                for expr in aggr_exprs {
+                    let mut aggr = try!(aggregate::build_aggr_func(expr));
+                    let args = box_try!(self.eval.batch_eval(expr.get_children()));
+                    try!(aggr.update(&args));
+                    aggrs.push(aggr);
+                }
+                self.gks.push(gk);
+                e.insert(aggrs);
+            }
+        }
+        Ok(())
+    }
+
+    /// Convert aggregate partial result to rows.
+    /// Data layout example:
+    /// SQL:	select count(c1), sum(c2) from t;
+    /// Aggs:	count(c1), sum(c2)
+    /// Rows:	groupKey1, count1, value1, count2, value2
```

Comment:
@shenli, can we change the format to only contains one value for each expression and two values only for avg?